cpp/src/io/json/nested_json_gpu.cu

```diff
@@ -0,0 +1,1293 @@
+                                       tokenizer_pda::get_translation_table(),
+                                       stream};
+
+  // Perform a PDA-transducer pass
+  json_to_tokens_fst.Transduce(d_pda_sgids.begin(),
+                               static_cast<SymbolOffsetT>(d_json_in.size()),
+                               d_tokens,
+                               d_tokens_indices,
+                               d_num_written_tokens,
+                               tokenizer_pda::start_state,
+                               stream);
+}
+
+std::pair<std::unique_ptr<column>, std::vector<column_name_info>> json_column_to_cudf_column(
+  json_column const& json_col, device_span<SymbolT const> d_input, rmm::cuda_stream_view stream)
+{
+  auto make_validity = [](json_column const& json_col) -> std::pair<rmm::device_buffer, size_type> {
+    if (json_col.current_offset == json_col.valid_count) { return {rmm::device_buffer{}, 0}; }
+
+    thrust::device_vector<json_column::row_offset_t> d_validity(
+      json_col.validity);  // TODO device_uvector.
+    return cudf::detail::valid_if(d_validity.begin(), d_validity.end(), thrust::identity<bool>{});
+  };
+
+  switch (json_col.type) {
+    case json_col_t::StringColumn: {
+      // move string_offsets to GPU and transform to string column
+      auto col_size            = json_col.string_offsets.size();
+      using char_length_pair_t = thrust::pair<const char*, size_type>;
+      CUDF_EXPECTS(json_col.string_offsets.size() == json_col.string_lengths.size(),
+                   "string offset, string length mismatch");
+      rmm::device_uvector<char_length_pair_t> d_string_data(col_size, stream);
+      rmm::device_uvector<json_column::row_offset_t> d_string_offsets =
+        cudf::detail::make_device_uvector_async(json_col.string_offsets, stream);
+      rmm::device_uvector<json_column::row_offset_t> d_string_lengths =
+        cudf::detail::make_device_uvector_async(json_col.string_lengths, stream);
+      auto offset_length_it =
+        thrust::make_zip_iterator(d_string_offsets.begin(), d_string_lengths.begin());
+      thrust::transform(rmm::exec_policy(stream),
+                        offset_length_it,
+                        offset_length_it + col_size,
+                        d_string_data.data(),
+                        [data = d_input.data()] __device__(auto ip) {
+                          return char_length_pair_t{data + thrust::get<0>(ip), thrust::get<1>(ip)};
+                        });
+      auto str_col_ptr                  = make_strings_column(d_string_data, stream);
+      auto [result_bitmask, null_count] = make_validity(json_col);
+      str_col_ptr->set_null_mask(result_bitmask, null_count);
+      return {std::move(str_col_ptr), {{"offsets"}, {"chars"}}};
+      break;
+    }
+    case json_col_t::StructColumn: {
+      std::vector<std::unique_ptr<column>> child_columns;
+      std::vector<column_name_info> column_names{};
+      size_type num_rows{-1};
```

Comment:
What's going on with this variable?  It's set multiple times in the loop from different child columns and then used at the end.  I assume the expectation is that all the child columns will simply have the same size.  Maybe something like this would be a little safer:

```
if(num_rows < 0){
   num_rows = child_column->size();
} else {
   CUDF_EXPECTS(num_rows == child_column->size());
}
```